### PR TITLE
Target label

### PR DESCRIPTION
**Pull request by antonrohr**
This makes the target label the same thing in the ProfilingTargetDialog and the Orbitmainwindow (and adds tests)

### CODE REVIEW

File: src/OrbitQt/CMakeLists.txt
Comment (beckerhe) [inline src/OrbitQt/CMakeLists.txt:119]: What has changed here?

File: src/OrbitQt/TargetLabel.h
Comment (beckerhe) [inline src/OrbitQt/TargetLabel.h:15]: I think you shouldn't inherit from QLabel here. To me it looks like the label should be a private member instead.

You're trying to enforce invariants by setting the label's properties in the setters of `TargetLabel`. But if you inherit from QLabel, the user of this class will be able to mess with this.

I see two solutions:
- Either make it a private member, so you have full control over the invariance and the user can rely on that.
- Or make it obvious that it's the user's responsibility, by making the setters free functions which take a label as an argument. `SetFileTarget(QLabel* target_label, const std::filesystem::path& path);`

Obviously the first is closer to the current solution.

File: src/OrbitQt/TargetLabelTest.cpp
Comment (beckerhe) [inline src/OrbitQt/TargetLabelTest.cpp:20]: I'm not saying you should remove it, only to think about it if this couldn't be checked differently.

It looks a lot like an implementation detail to me. In the end for the user of the class it should be of no concern
how the font coloring is done. You could also use QPalette, or overwrite the paintEvent method, or even role your own label implementation. The only important thing in the end is the color of the text, not the stylesheet.

And by putting that check into the tests you even guarantee to the user that this is officially observable behaviour. That means someone could start checking the stylesheet property to figure out in which mode the target label currently is. Still unlikely for this use-case, but I hope you get the idea.

File: src/OrbitQt/TargetLabelTest.cpp
Comment (reichlfl) [inline src/OrbitQt/TargetLabelTest.cpp:103]: Nit: Since you test this twice (here and in `SetProcessAlive`), it would be easier to maintain and update if you'd move it to a function. Same for the three tests below.

File: src/OrbitQt/TargetLabel.cpp
Comment (beckerhe) [inline src/OrbitQt/TargetLabel.cpp:37]: Nit: It's a sentence, I would make a fullstop at the end.

File: src/OrbitQt/TargetLabel.h
Comment (reichlfl) [inline src/OrbitQt/TargetLabel.h:20]: I'm not super happy with the API here - there's 3 methods that look like regular setters, but their usage is mutually exclusive.
Maybe just rename them to make this clear, e.g. `ChangeToXXXTarget`.

File: src/OrbitQt/TargetLabel.h
Comment (reichlfl) [inline src/OrbitQt/TargetLabel.h:23]: Same as above. Maybe `SetStatus` with an enum parameter, and an optional tooltip parameter?

File: src/OrbitQt/TargetLabel.h
Comment (reichlfl) [inline src/OrbitQt/TargetLabel.h:28]: Nit: These should be private, and I assume they are public because of the unit test? I'd avoid making something public just to be testable - it's good practice to test through the public API and ignore internals, so this indicates you're testing internals.

However - since this whole widget has purely visual purposes, this level of detail may be fine here... I just wanted to point this out, but I'm undecided.

**Review by reichlfl — COMMENTED** [review]
This is nice, the encapsulation makes it a lot cleaner. PTAL at my comments regarding the API, I'd suggest to at least rename the methods to sound less like setters.

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

File: src/OrbitQt/TargetLabel.h
Comment (reichlfl) [reply src/OrbitQt/TargetLabel.h:15]: I second this, because it'll also make it easier to extend it by e.g. adding a "Connected" icon as we discussed.

File: src/OrbitQt/CMakeLists.txt
Comment (antonrohr) [reply src/OrbitQt/CMakeLists.txt:119]: Took me a while to figure it out. I think vscode or clang format fixed line endings here (CRLF -> LF). 
Screenshot 2021-03-11 at 10 49 37

Comment (beckerhe) [reply src/OrbitQt/CMakeLists.txt:119]: So nothing changed and it can be removed from this commit?

Comment (antonrohr) [reply src/OrbitQt/CMakeLists.txt:119]: Well. The file contained mixed line endings before this commit. Which is a bad thing in the first place. Now its just LF line endings and thats a good thing. I could add a dedicated "Fix line endings" commit, but for the sake of simplicity, I would leave it as is. wdyt?

Comment (beckerhe) [reply src/OrbitQt/CMakeLists.txt:119]: I personally value clean commits a lot higher than consistent line endings, but I also don't feel strongly at all about this. So if you prefer to leave it, that's fine with me!

File: src/OrbitQt/TargetLabelTest.cpp
Comment (antonrohr) [reply src/OrbitQt/TargetLabelTest.cpp]: outdated

File: src/OrbitQt/TargetLabel.h
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.h]: I changed it, ptal

File: src/OrbitQt/TargetLabel.h
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.h]: I changed it, ptal

File: src/OrbitQt/TargetLabel.h
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.h]: I made them an implementation detail and dont test for correct color anymore, only for change in color. ptal

File: src/OrbitQt/TargetLabel.h
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.h]: I implemented the first solution (in a way). I mark this as resolved

File: src/OrbitQt/TargetLabelTest.cpp
Comment (antonrohr) [reply src/OrbitQt/TargetLabelTest.cpp]: As we talked offline, I changed the testing of this. It now uses QPalette instead of spreadsheet and I only test for change in color.

File: src/OrbitQt/TargetLabel.cpp
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.cpp]: done

**Comment by antonrohr** [comment]
I changed quite a bit in this PR and I needed the cmake change that was submitted yesterday, hence I rebased and force pushed. @reichlfl @beckerhe please take a look again. 

This now also adds cpu utilization and icons and therefore does everything tracked in http://b/177305224

File: src/OrbitQt/ProfilingTargetDialog.cpp
Comment (beckerhe) [inline src/OrbitQt/ProfilingTargetDialog.cpp:274]: why not
QObject::connect(&state_stadia_process_selected_, &QState::exited, ui_->targetLabel, &TargetLabel::Clear);?

File: src/OrbitQt/ProfilingTargetDialog.cpp
Comment (beckerhe) [inline src/OrbitQt/ProfilingTargetDialog.cpp:470]: The right hand side of this comparison is constant, isn't it? It's hard to understand at first that this doesn't change. I'd suggest to put this pid into a variable first, and then search for the correct process by comparing to this variable.

Comment (beckerhe) [reply src/OrbitQt/ProfilingTargetDialog.cpp:470]: One more thing: The "current index" is somehow independent from a selection, and mainly meant for keyboard navigation.

You can have a selection and a different current index, or maybe no current index at all. I think you wanna ask the selection model for the selectedRows() and then take the first one.

File: src/OrbitQt/orbitmainwindow.cpp
Comment (beckerhe) [inline src/OrbitQt/orbitmainwindow.cpp:1177]: Nit: This feels a bit like a dependency on the implemenation of TargetLabel, because you need to know which method of TargetLabel changes the widget's size. Ideally this is decoupled and TargetLabel tells you when a call of adjustSize is needed.

This could be done by a signal in TargetLabel. Maybe there is already one in QWidget which could be used. Haven't checked though.

**Review by beckerhe — COMMENTED** [review]
Very Nice, Anton. I added some comments

File: src/OrbitQt/TargetLabel.h
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.h:59]: They are actually not there because of the tests, but because I "compute" them. Aka take the white icon and color it pixel by pixel (see TargetDialog.cpp anon namespace `ColorizeIcon()`). I did not want to make this computation everytime I use them, but once. I test via the enum. 

The static is a good idea. I will change it to static

Comment (beckerhe) [reply src/OrbitQt/TargetLabel.h:59]: But that still doesn't explain why they need to be in the class. Why not static variables in the cpp file? You could still compute them.

File: src/OrbitQt/orbitmainwindow.cpp
Comment (antonrohr) [reply src/OrbitQt/orbitmainwindow.cpp]: I like this idea a lot. QWidget did not have a signal that was useful, so I added a `SizeChanged` signal. done

File: src/OrbitQt/ProfilingTargetDialog.cpp
Comment (antonrohr) [reply src/OrbitQt/ProfilingTargetDialog.cpp]: done

File: src/OrbitQt/ProfilingTargetDialog.cpp
Comment (antonrohr) [reply src/OrbitQt/ProfilingTargetDialog.cpp]: done

File: src/OrbitQt/ProfilingTargetDialog.cpp
Comment (antonrohr) [reply src/OrbitQt/ProfilingTargetDialog.cpp]: done

File: src/OrbitQt/ProfilingTargetDialog.cpp
Comment (antonrohr) [reply src/OrbitQt/ProfilingTargetDialog.cpp]: Thank you! Missed that. done

File: src/OrbitQt/TargetLabel.h
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.h]: done

File: src/OrbitQt/TargetLabel.cpp
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.cpp]: Did not know about that arg overload. neat! done

File: src/OrbitQt/TargetLabel.cpp
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.cpp]: done

File: src/OrbitQt/TargetLabel.h
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.h]: I renamed it to IconType. For easy reviewing its a separate fixup rename commit. ptal

File: src/OrbitQt/TargetLabel.h
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.h]: done

File: src/OrbitQt/TargetLabel.h
Comment (antonrohr) [reply src/OrbitQt/TargetLabel.h]: done

File: src/OrbitQt/orbitmainwindow.cpp
Comment (beckerhe) [inline src/OrbitQt/orbitmainwindow.cpp:9]: Nit

File: src/OrbitQt/TargetLabel.cpp
Comment (beckerhe) [inline src/OrbitQt/TargetLabel.cpp:8]: Nit

**Review by beckerhe — APPROVED** [review]
Thanks a lot for the fixes. LGTM

**Comment by antonrohr** [comment]
@reichlfl I think we are good here code wise with Hennings review, maybe you could just look at these screenshots to make sure we are on the same page.

The blurriness in the red color is an encoding artifact of chrome remote desktop. I dont know what the encoders problem with red color is, but its definitely sharp without CRD.

Screenshot 2021-03-18 at 10 56 00
Screenshot 2021-03-18 at 10 55 00
Screenshot 2021-03-18 at 10 55 20
Screenshot 2021-03-18 at 10 54 54
Screenshot 2021-03-18 at 10 54 42

File: src/OrbitQt/orbitmainwindow.cpp
Comment (antonrohr) [reply src/OrbitQt/orbitmainwindow.cpp]: One day I will find time to setup auto fixing iwyu on every save. This gets really annoying. Thank you for pointing it out. done (and rebased)

File: src/OrbitQt/TargetLabel.h
Comment (beckerhe) [reply src/OrbitQt/TargetLabel.h]: Sounds good.

**Comment by reichlfl** [comment]
> @reichlfl I think we are good here code wise with Hennings review, maybe you could just look at these screenshots to make sure we are on the same page.

Sorry, I didn't realize I had open discussion items here. I'll trust in your and Hennings judgment on this, no need to wait for me to do a second review.

Regarding the screenshots: Really nice, I like the inclusion of the icon. A nitpick, the red looks really aggressive to a point where it's a bit hard to read on the grey background, maybe it should be a bit of a lighter shade. But that might also be due to the compression.

Orange indicates the process connection was lost, but the instance is still alive? A different icon for this could be nice, but not urgently needed. Could also be changed at a later point. Do we have tooltips that indicate the difference between the red and orange state? It's not clear from just looking at the label / icon.

Maybe we just use a "Warning" icon when the process connection is lost? https://material.io/resources/icons/?search=problem&icon=report_problem&style=outline
In the future, I can imagine you can still take a capture as we more and more become a system profiler (e.g. memory, scheduling and GPU information is still valid), so "Warning" may be appropriate.

**Review by reichlfl — APPROVED** [review]

**Comment by antonrohr** [comment]
Thanks for the suggestions Florian. I changed the icon and made the red a bit brighter. There are also tooltips. The process has ended one could be a bit more verbose, in case you have better idea. 

Screenshot 2021-03-18 at 13 31 58
Screenshot 2021-03-18 at 13 41 29
Capture

**Comment by reichlfl** [comment]
> Thanks for the suggestions Florian. I changed the icon and made the red a bit brighter. There are also tooltips. The process has ended one could be a bit more verbose, in case you have better idea.
> 

More nitpicking, I'd probably make the icons a bit smaller - but this is really a minor detail at this point.
Can we already automatically reconnect to the process? Then I'd state that in the tooltip (e.g. "Restart a process with a matching name to continue profiling").

**Comment by antonrohr** [comment]
I made the icons smaller. 

The reconnection to a restarted process is not yet implemented. The actual change is just a couple of lines, but I am somewhat afraid to break something at a different.  This will be a different PR. (I will remember to adjust the tooltip)

Screenshot 2021-03-18 at 16 33 25
Screenshot 2021-03-18 at 16 32 57

**Comment by reichlfl** [comment]
> I made the icons smaller.
> 
> The reconnection to a restarted process is not yet implemented. The actual change is just a couple of lines, but I am somewhat afraid to break something at a different. This will be a different PR. (I will remember to adjust the tooltip)

Makes sense, thanks!

**Pull request merged**
